lib/daru/dataframe.rb

```diff
@@ -2302,30 +2301,25 @@ def insert_or_modify_vector name, vector
 
           v = Daru::Vector.new(vector, name: set_name(name), index: @index)
         end
-
         assign_or_add_vector name, v
       end
     end
 
     def assign_or_add_vector name, v
-      #FIXME: fix this jugaad. need to make changes in Indexing itself.
-      begin
-        pos = @vectors[name]
-      rescue IndexError
-        pos = name
-      end
+      pos = @vectors[*name]
 
-      if !pos.kind_of?(Daru::Index) and pos == name and
-        (@vectors.include?(name) or (pos.is_a?(Integer) and pos < @data.size))
-        @data[pos] = v
-      elsif pos.kind_of?(Daru::Index)
-        pos.each do |p|
-          @data[@vectors[p]] = v
-        end
+      # Update an existing vector based on its index
+      if pos.is_a? Daru::Index
+        # If there are multiple vectors associated with this name
+        pos.each { |p| @data[@vectors[p]] = v }
       else
-        @vectors = @vectors | [name] if !@vectors.include?(name)
-        @data[@vectors[name]] = v
+        @data[pos] = v
       end
+
+    rescue IndexError
```

Comment:
I assume such problems suggest we need extend indexes API (so that there'll be an easy way to ask "if index includes this").
Also, I can't see much specs to cover this problem (variants of indexes).